ush/python/pygfs/jedi/jedi.py

```diff
@@ -0,0 +1,271 @@
+#!/usr/bin/env python3
+
+import os
+from logging import getLogger
+from typing import List, Dict, Any, Optional
+from jcb import render
+from wxflow import (AttrDict,
+                    chdir, rm_p,
+                    parse_j2yaml,
+                    logit,
+                    Task,
+                    Executable,
+                    WorkflowException)
+
+logger = getLogger(__name__.split('.')[-1])
+
+
+class JEDI:
+
+    def __init__(self, task_config: AttrDict[str, Any]) -> None:
+
+        # For provenance, save incoming task_config as a private attribute of JEDI object
+        self._task_config = task_config
+
+        _exe_name = os.path.basename(task_config.JEDIEXE)
+
+        self.exe = os.path.join(task_config.DATA, _exe_name)
+        self.yaml = os.path.join(task_config.DATA, os.path.splitext(_exe_name)[0] + '.yaml')
+        self.config = AttrDict()
+        self.j2tmpl_dir = os.path.join(task_config.PARMgfs, 'gdas')
+
+    @logit(logger)
+    def set_config(self, task_config: AttrDict[str, Any], algorithm: Optional[str] = None) -> AttrDict:
+        """Compile a JEDI configuration dictionary from a template file and save to a YAML file
+
+        Parameters
+        ----------
+        task_config : AttrDict
+            Dictionary of all configuration variables associated with a GDAS task.
+        algorithm (optional) : str
+            Name of the algorithm used to generate the JEDI configuration dictionary.
+            It will override the algorithm set in the task_config.JCB_<>_YAML file.
+
+        Returns
+        ----------
+        None
+        """
+
+        if 'JCB_BASE_YAML' in task_config.keys():
+            # Step 1: Fill templates of the JCB base YAML file
+            jcb_config = parse_j2yaml(task_config.JCB_BASE_YAML, task_config)
+
+            # Step 2: If algorithm is present then override the algorithm in the JEDI
+            #         config. Otherwise, if the algorithm J2-YAML is present, fill
+            #         its templates and merge.
+            if algorithm:
+                jcb_config['algorithm'] = algorithm
+            elif 'JCB_ALGO' in task_config.keys():
+                jcb_config['algorithm'] = task_config.JCB_ALGO
+            elif 'JCB_ALGO_YAML' in task_config.keys():
+                jcb_algo_config = parse_j2yaml(task_config.JCB_ALGO_YAML, task_config)
+                jcb_config = {**jcb_config, **jcb_algo_config}
+
+            # Step 3: Generate the JEDI YAML using JCB
+            self.config = render(jcb_config)
+        elif 'JEDIYAML' in task_config.keys():
+            # Generate JEDI YAML without using JCB
+            self.config = parse_j2yaml(task_config.JEDIYAML, task_config,
```

Comment:
This is a nit-pick comment, but `config` has a very generic implication.  Would `jedi_config` be acceptable to  make it clear that this `jedi_config` is a result of rendering `JEDIYAML`

Reply:
In the case of ```self``` it's not clear, but when the object is called ```jedi```, it feels redundant. But I see your point. In that case, would you want ```jedi.exe``` and ```jedi.yaml``` to become ```jedi.jedi_exe``` and ```jedi.jedi_yaml``` as well?

Reply:
Good point.  I am fine with leaving out the `jedi_` prefixes for `jedi` objects.